src/lib/holocene/api-pagination.svelte

```diff
@@ -1,73 +1,127 @@
   import { options } from '$lib/stores/pagination';
+  import { onMount } from 'svelte';
 
   type T = $$Generic;
   type PaginatedRequest = (
     size: number,
-    token: string,
-  ) => Promise<{ items: any[]; nextPageToken: string }>;
+    token: NextPageToken,
+  ) => Promise<{ items: any[]; nextPageToken: NextPageToken }>;
 
   export let onError: (error: any) => void;
   export let onFetch: () => Promise<PaginatedRequest>;
+  export let onShiftUp: (event: KeyboardEvent) => void | undefined = undefined;
+  export let onShiftDown: (event: KeyboardEvent) => void | undefined =
+    undefined;
+  export let onSpace: (event: KeyboardEvent) => void | undefined = undefined;
 
-  let store = createPaginationStore();
-  let error: any;
+  export let pageSizeOptions: string[] = options;
+  export let total: string | number = '';
 
-  $: nextIndex = $store.nextIndex;
-  $: pageSize = $store.pageSize;
+  let store = createPaginationStore(pageSizeOptions);
+  let error: any;
+  let shifted = false;
 
   function clearError() {
     if (error) error = undefined;
   }
 
-  async function onPageChange(nextIndex: number) {
-    clearError();
-    store.setUpdating();
-    try {
-      const fetchData: PaginatedRequest = await onFetch();
-      const response = await fetchData(pageSize, $store.indexTokens[nextIndex]);
-      const { items, nextPageToken } = response;
-      store.setNextPageToken(nextPageToken, items);
-    } catch (err: any) {
-      error = err;
-      onError(error);
-    }
+  onMount(() => {
+    initalDataFetch();
+  });
+
+  async function initalDataFetch() {
+    const fetchData: PaginatedRequest = await onFetch();
+    const response = await fetchData($store.pageSize, '');
+    const { items, nextPageToken } = response;
+    store.nextPageWithItems(nextPageToken, items);
   }
 
-  async function onPageSizeChange(pageSize: number) {
+  async function fetchIndexData() {
     clearError();
-    store.resetPageSize();
-    try {
-      const fetchData: PaginatedRequest = await onFetch();
-      const response = await fetchData(
-        pageSize,
-        $store.indexTokens[$store.index],
-      );
-      const { items, nextPageToken } = response;
-      store.setNextPageToken(nextPageToken, items);
-    } catch (err: any) {
-      error = err;
-      onError(error);
+    store.setUpdating();
+    if (!$store.hasNextIndexData) {
+      try {
+        const fetchData: PaginatedRequest = await onFetch();
+        const response = await fetchData(
+          $store.pageSize,
+          $store.indexData[$store.index].nextToken,
+        );
+        const { items, nextPageToken } = response;
+        store.nextPageWithItems(nextPageToken, items);
+      } catch (err: any) {
+        error = err;
+        onError(error);
+      }
+    } else {
+      store.nextPage();
     }
   }
 
-  $: {
-    onPageChange(nextIndex);
+  async function handleKeydown(event: KeyboardEvent) {
```

Comment:
Do we have unit tests for this? I ask because I don't see it exported. It's kind of weird that it's in a `.svelte` file at all. I'd rather see us use dependency injection and pull this out completely. This is kind of like what I said to @rossedfort in I think the signal PR.